### PR TITLE
GEO: add GEOADD API for redis proxy to support full GEO function

### PR DESCRIPTION
**Pull request by acelyc111**
### What problem does this PR solve? 
add GEOADD API for redis proxy to support full GEO function

### What is changed and how it works?
add an API for redis proxy: GEOADD

#### Tests

- Unit test
Yes
- Integration test
Yes
- Manual test (add detailed scripts or steps below)
Yes
- No code
No

#### Code changes

- Has exported function/method change
No
- Has exported variable/fields change
No
- Has interface methods change
No
- Has persistent data change
No

#### Side effects

- Possible performance regression
No
- Increased code complexity
A little
- Breaking backward compatibility
No

#### Related changes

- Need to cherry-pick to the release branch
Yes
- Need to update the documentation
Yes
- Need to be included in the release note
Yes

### CODE REVIEW

File: src/geo/bench/bench.cpp [hunk 1]
```diff
@@ -65,14 +65,12 @@ int main(int argc, char **argv)
...
         for (int i = 0; i < data_count; ++i) {
+            std::string value;
             S2LatLng latlng(S2Testing::SamplePoint(rect));
-            std::string id = std::to_string(i);
-            std::string value = id + "|2018-06-05 12:00:00|2018-06-05 13:00:00|abcdefg|" +
-                                std::to_string(latlng.lng().degrees()) + "|" +
-                                std::to_string(latlng.lat().degrees()) + "|123.456|456.789|0|-1";
-
-            int ret = my_geo.set(id, "", value, 1000);
+            assert(codec.encode_to_value(latlng.lat().degrees(), latlng.lng().degrees(), value));
```
Comment (qinzuoyan) [inline src/geo/bench/bench.cpp]: 不能把有效的逻辑代码放在assert里面，如果哪天关闭了assert宏，这个代码就执行不到了

File: src/geo/lib/latlng_codec.cpp [hunk 2]
```diff
@@ -59,17 +62,44 @@ bool latlng_extractor::extract_from_value(const std::string &value, S2LatLng &la
...
+    if (!latlng.is_valid()) {
+        derror_f("latlng is invalid. lat_degrees={}, lng_degrees={}", lat_degrees, lng_degrees);
+        return false;
+    }
+
+    value.clear();
+    int index = 0;
+    for (int i = 0; i <= _sorted_indices[1]; ++i) {
+        if (i == _sorted_indices[index]) {
+            if ((index == 0 && _latlng_order) || (index == 1 && !_latlng_order)) {
```
Comment (qinzuoyan) [inline src/geo/lib/latlng_codec.cpp]: 如果{}只有单行代码，其实可以不用{}，显得比较简洁

File: src/geo/test/latlng_codec_test.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    ASSERT_TRUE(codec.encode_to_value(lat_degrees, lng_degrees, value));
+    ASSERT_EQ("|67.890000||12.345000", value);
+
+    ASSERT_FALSE(codec.encode_to_value(91, lng_degrees, value));
+    ASSERT_FALSE(codec.encode_to_value(-91, lng_degrees, value));
+    ASSERT_FALSE(codec.encode_to_value(lat_degrees, 181, value));
+    ASSERT_FALSE(codec.encode_to_value(lat_degrees, -181, value));
+}
+} // namespace geo
+} // namespace pegasus
```
Comment (levy5307) [inline src/geo/test/latlng_codec_test.cpp]: 少最后的空行

Comment (acelyc111) [reply src/geo/test/latlng_codec_test.cpp]: 这里是有的,github会自动隐藏. 如果没有文件末尾空行, github会有提示的

File: src/geo/lib/latlng_codec.cpp (hunk 2 above)
Comment (acelyc111) [reply src/geo/lib/latlng_codec.cpp]: 按Google code style guide倒没有强制要求, 两种都可以, 需要保证同一个if..else..的风格一致即可

Comment (neverchanje) [reply src/geo/lib/latlng_codec.cpp]: 相反的，cpplint/clang-tidy 要求任何 if {} else {} 都要用花括号标起来：https://clang.llvm.org/extra/clang-tidy/checks/readability-braces-around-statements.html。
因为有的程序员没看清楚，容易写错，像这种：

```
if (enabled)
   a++;
b++;
```

Comment (acelyc111) [reply src/geo/lib/latlng_codec.cpp]: > 相反的，cpplint/clang-tidy 要求任何 if {} else {} 都要用花括号标起来：https://clang.llvm.org/extra/clang-tidy/checks/readability-braces-around-statements.html
> 因为有的程序员没看清楚，容易写错，像这种：
> 
> ```
> if (enabled)
>    a++;
> b++;
> ```

我个人也习惯于都括起来

Comment (qinzuoyan) [reply src/geo/lib/latlng_codec.cpp]: ok吧

**Pull request merged**